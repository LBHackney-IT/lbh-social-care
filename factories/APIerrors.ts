import { Factory } from 'fishery';

import { ErrorAPI } from 'types';

export const errorFactory = Factory.define<ErrorAPI>(() => ({
  response: {
    data: {},
    status: 404,
    statusText: 'error',
    headers: {},
    config: {},
  },
  name: 'error',
  message: 'error',
  config: {},
  isAxiosError: false,
  toJSON: () => ({}),
}));

export const mockedAPIerror = errorFactory.build();
export const mockedAPIservererror = errorFactory.build({
  response: {
    status: 500,
  },
});
