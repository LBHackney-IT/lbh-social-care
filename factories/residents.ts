import { Factory } from 'fishery';

import { Resident, ExtendedResident } from 'types';

export const legacyResidentFactory = Factory.define<Resident>(
  ({ sequence }) => ({
    mosaicId: sequence,
    dateOfBirth: '2020-11-13',
    firstName: 'Foo',
    lastName: 'Bar',
    nhsNumber: '12345',
    ageContext: 'A',
    gender: 'F',
  })
);

export const residentFactory = Factory.define<ExtendedResident>(
  ({ sequence }) => ({
    personId: sequence,
    dateOfBirth: '2020-11-13',
    firstName: 'Foo',
    lastName: 'Bar',
    nhsNumber: 12345,
    contextFlag: 'A',
    gender: 'F',
    createdBy: 'foo@bar.com',
    otherNames: [],
    phoneNumbers: [],
  })
);

export const mockedLegacyResident = legacyResidentFactory.build();
export const mockedResident = residentFactory.build();
