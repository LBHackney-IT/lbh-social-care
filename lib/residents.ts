import axios from 'axios';

import type { LegacyResident, Resident, ResidentsAPI, AgeContext } from 'types';

const { ENDPOINT_API, AWS_KEY } = process.env;

const headers = { 'x-api-key': AWS_KEY };

interface ResidentBE extends Omit<LegacyResident, 'restricted'> {
  restricted: 'Y' | 'N';
  addressList: Array<{
    displayAddressFlag: 'Y' | 'N';
    addressLine1: string;
    postCode: string;
    uprn: string;
  }>;
}

const sanitiseResidentData = (residents: ResidentBE[]): LegacyResident[] =>
  residents?.map(
    ({ addressList, ...resident }: ResidentBE): LegacyResident => {
      const address = addressList?.find(
        ({ displayAddressFlag }) => displayAddressFlag === 'Y'
      );
      return {
        ...resident,
        restricted: resident.restricted === 'Y',
        address: address && {
          address: address.addressLine1,
          postcode: address.postCode,
          uprn: address.uprn,
        },
      };
    }
  );

export const getResidents = async (
  params: Record<string, unknown>
): Promise<ResidentsAPI> => {
  const { data } = await axios.get(`${ENDPOINT_API}/residents`, {
    headers,
    params,
  });
  return { ...data, residents: sanitiseResidentData(data.residents) };
};

export const getResident = async (
  personId: number,
  params: { context_flag: AgeContext }
): Promise<Resident> => {
  const { data } = await axios.get(`${ENDPOINT_API}/residents/${personId}`, {
    headers,
    params,
  });
  return { ...data, restricted: data.restricted === 'Y' };
};

export const normalisePhoneInput = (formData: {
  phoneNumbers?: Record<string, string>[];
  [key: string]: unknown;
}): Record<string, unknown> => ({
  ...formData,
  phoneNumbers: formData.phoneNumbers
    ? formData.phoneNumbers.map((phone) => ({
        ...phone,
        type: phone.type === '' ? 'main' : phone.type,
      }))
    : null,
});

export const addResident = async (
  formData: Record<string, unknown>
): Promise<ResidentsAPI> => {
  const { data } = await axios.post(
    `${ENDPOINT_API}/residents`,
    normalisePhoneInput(formData),
    {
      headers: { ...headers, 'Content-Type': 'application/json' },
    }
  );
  return data;
};

export const updateResident = async (
  formData: Record<string, unknown>
): Promise<ResidentsAPI> => {
  const { data } = await axios.patch(
    `${ENDPOINT_API}/residents`,
    normalisePhoneInput(formData),
    {
      headers: { ...headers, 'Content-Type': 'application/json' },
    }
  );
  return data;
};
