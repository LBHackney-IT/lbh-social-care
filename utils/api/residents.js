import axios from 'axios';

export const getResidents = async (params) => {
  const { data } = await axios.get('/api/residents', {
    params,
  });
  return data;
};

export const getResident = async (id, params) => {
  const { data } = await axios.get(`/api/residents/${id}`, {
    params,
  });
  return data;
};

export const addResident = async (formData) => {
  const { data } = await axios.post(`/api/residents`, formData);
  return data;
};
