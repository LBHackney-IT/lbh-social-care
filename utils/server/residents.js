import axios from 'axios';

const {
  ENDPOINT_API,
  ENDPOINT_MOSAIC,
  AWS_KEY,
  AWS_AUTHORIZATION,
} = process.env;

export const getResidents = async (params) => {
  const { data } = await axios.get(`${ENDPOINT_API}/residents`, {
    headers: {
      'x-api-key': AWS_KEY,
    },
    params,
  });
  return data.residents;
};

export const getResident = async (id) => {
  const { data } = await axios.get(`${ENDPOINT_MOSAIC}/residents/${id}`, {
    headers: {
      Authorization: AWS_AUTHORIZATION,
    },
  });
  return data;
};
