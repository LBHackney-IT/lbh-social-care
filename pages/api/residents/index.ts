import { StatusCodes } from 'http-status-codes';

import { getResidents, addResident } from 'lib/residents';
import { isAuthorised } from 'utils/auth';

import type { NextApiRequest, NextApiResponse, NextApiHandler } from 'next';

const endpoint: NextApiHandler = async (
  req: NextApiRequest,
  res: NextApiResponse
) => {
  const user = isAuthorised(req);
  if (!user) {
    return res.status(StatusCodes.UNAUTHORIZED).end();
  }
  if (!user.isAuthorised) {
    return res.status(StatusCodes.FORBIDDEN).end();
  }
  switch (req.method) {
    case 'GET':
      try {
        const data = await getResidents({
          ...req.query,
          context_flag: user.permissionFlag,
        });
        res.status(StatusCodes.OK).json(data);
      } catch (error) {
        console.error('Residents get error:', error?.response?.data);
        error?.response?.status === StatusCodes.NOT_FOUND
          ? res
              .status(StatusCodes.NOT_FOUND)
              .json({ message: 'Residents Not Found' })
          : res
              .status(StatusCodes.INTERNAL_SERVER_ERROR)
              .json({ message: 'Unable to get the Residents' });
      }
      break;

    case 'POST':
      try {
        const data = await addResident(req.body);
        res.status(StatusCodes.OK).json(data);
      } catch (error) {
        console.error('Resident post error:', error?.response?.data);
        console.error('Resident post request:', req);
        res
          .status(StatusCodes.INTERNAL_SERVER_ERROR)
          .json({ message: 'Unable to add resident' });
      }
      break;

    default:
      res
        .status(StatusCodes.BAD_REQUEST)
        .json({ message: 'Invalid request method' });
  }
};

export default endpoint;
