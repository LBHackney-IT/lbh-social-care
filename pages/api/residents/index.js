import * as HttpStatus from 'http-status-codes';

import { getResidents, addResident } from 'utils/server/residents';
import { isAuthorised } from 'utils/auth';

export default async (req, res) => {
  if (!isAuthorised(req)) {
    return res
      .status(HttpStatus.UNAUTHORIZED)
      .json({ message: 'Auth cookie missing.' });
  }
  switch (req.method) {
    case 'GET':
      try {
        const data = await getResidents(req.query);
        res.status(HttpStatus.OK).json(data);
      } catch (error) {
        console.log('Residents get error:', error?.response?.data);
        error?.response?.status === HttpStatus.NOT_FOUND
          ? res
              .status(HttpStatus.NOT_FOUND)
              .json({ message: 'Residents Not Found' })
          : res
              .status(HttpStatus.INTERNAL_SERVER_ERROR)
              .json({ message: 'Unable to get the Residents' });
      }
      break;

    case 'POST':
      try {
        const data = await addResident(req.body);
        res.status(HttpStatus.OK).json(data);
      } catch (error) {
        console.log('Resident post error:', error?.response?.data);
        res
          .status(HttpStatus.INTERNAL_SERVER_ERROR)
          .json({ message: 'Unable to add resident' });
      }
      break;

    default:
      res
        .status(HttpStatus.BAD_REQUEST)
        .json({ message: 'Invalid request method' });
  }
};
