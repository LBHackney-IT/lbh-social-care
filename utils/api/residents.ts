import axios from 'axios';
import useSWR, {
  useSWRInfinite,
  SWRResponse,
  SWRInfiniteSWRResponse,
} from 'swr';

import { getInfiniteKey } from 'utils/api';

import type { Resident, ErrorAPI } from 'types';

export const useResidents = (
  params: Record<string, unknown>,
  invoke = true
): SWRInfiniteSWRResponse<Record<string, unknown>, Error> =>
  useSWRInfinite(
    // @ts-ignore
    invoke ? getInfiniteKey('/api/residents', 'residents', params) : null
  );

export const useResident = (
  personId: number
): SWRResponse<Resident, ErrorAPI> => useSWR(`/api/residents/${personId}`);

export const addResident = async (
  formData: Record<string, unknown>
): Promise<Record<string, unknown>> => {
  const { data } = await axios.post(`/api/residents`, formData);
  return { ref: data?.personId, data };
};

export const updateResident = async (
  formData: Record<string, unknown>
): Promise<Record<string, unknown>> => {
  const { data } = await axios.patch(`/api/residents`, formData);
  return { ref: data?.personId, data };
};
