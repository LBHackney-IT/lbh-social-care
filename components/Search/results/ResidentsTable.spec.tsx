import { render } from '@testing-library/react';

import ResidenTable from './ResidentsTable';
import { mockedLegacyResident } from 'factories/residents';

describe('ResidenTable component', () => {
  const props = {
    records: [mockedLegacyResident],
  };
  it('should render properly', () => {
    const { asFragment } = render(<ResidenTable {...props} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
