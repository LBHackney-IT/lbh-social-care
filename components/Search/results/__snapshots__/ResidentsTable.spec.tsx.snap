// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResidenTable component should render properly 1`] = `
<DocumentFragment>
  <table
    class="govuk-table"
  >
    <thead
      class="govuk-table__head"
    >
      <tr
        class="govuk-table__row"
      >
        <th
          class="govuk-table__header"
          scope="col"
        >
          Mosaic ID
        </th>
        <th
          class="govuk-table__header"
          scope="col"
        >
          Client Name
        </th>
        <th
          class="govuk-table__header"
          scope="col"
        >
          Date of birth
        </th>
        <th
          class="govuk-table__header"
          scope="col"
        />
      </tr>
    </thead>
    <tbody
      class="govuk-table__body"
    >
      <tr
        class="govuk-table__row"
      >
        <td
          class="govuk-table__cell"
        >
          1
        </td>
        <td
          class="govuk-table__cell"
        >
          Foo Bar
        </td>
        <td
          class="govuk-table__cell"
        >
          13/11/2020
        </td>
        <td
          class="govuk-table__cell"
        >
          <a
            class="govuk-link govuk-custom-text-color"
            href="/people/1"
          >
            View
          </a>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
